potassco/graph.h

```diff
@@ -0,0 +1,166 @@
+// {{{ MIT License
+
+// Copyright 2017 Roland Kaminski
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+
+// }}}
+
+
+#pragma once
+
+#include <forward_list>
+#include <vector>
+
+namespace Potassco {
+/////////////////////////////////////////////////////////////////////////////////////////
+// Graph template
+/////////////////////////////////////////////////////////////////////////////////////////
+template <class T>
+class Graph {
+public:
+    struct Node;
+    using NodeVec = std::vector<Node*>;
+    using SCCVec  = std::vector<NodeVec>;
+
+    struct Node {
+        friend class Graph;
+        template <class... U>
+        Node(unsigned phase, U&& ...args);
+
+        Node(const Node& other) = delete;
+        Node(Node&& other) noexcept = default;
+        Node& operator=(const Node& other) = delete;
+        Node& operator=(Node&& other) noexcept = default;
+        ~Node() noexcept = default;
+
+        void insertEdge(Node& n);
+        auto begin() const -> typename NodeVec::const_iterator;
+        auto end()   const -> typename NodeVec::const_iterator;
+
+        // NOLINTNEXTLINE
+        T data;
+
+    private:
+        NodeVec edges_;
+        unsigned visited_;
+        typename NodeVec::iterator finished_;
+    };
+
+    Graph() = default;
+    Graph(Graph&& other) noexcept = default;
+    Graph(const Graph& other) = delete;
+    Graph& operator=(Graph&& other) noexcept = default;
+    Graph& operator=(const Graph&) = delete;
+    ~Graph() = default;
+
+    auto tarjan() -> SCCVec;
+
+    template <class... U>
+    auto insertNode(U&& ...x) -> Node&;
+
+private:
+    unsigned nphase() { return phase_ == 0 ? 1 : 0; }
+    using NodeList = std::forward_list<Node>;
```

Comment:
What's the motivation for using a `forward_list` here? It implies a "per-node" allocation and an additional link "per-node". Why not use a `std::vector<std::unique_ptr<Node>>`? While this is also bad in terms of number allocations and node locality, it at least avoids the additional link and also has trivial  traversal code.